### ISSUE TITLE
Query is slow when using "inside"

### ISSUE THREAD

**Issue by johanPCQLMB** [JanusGraph/janusgraph#1074]
Hello, 

I am using the configuration defined on here (default one for gremlin server) : http://docs.janusgraph.org/0.2.0/server.html with cassandra and elasticsearch. 
I am experiencing a performance issue when I query the database using the "inside" function. 
I have created a graph that contains about 30 million vertices around 6 millions of which have the property "beginning". 
I am looking to get the vertices with the "beginning" value between a value x and y. The query should return several thousands of vertices, each vertex has 4 properties.

I have created an index in my code as follows : 

//init janusGraphManagement Object 
var mgmt: JanusGraphManagement = graph.openManagement()
//defining label of my vertices
val label = mgmt.makeVertexLabel("Journey").make()
//defining the beginning value
val beginning = mgmt.makePropertyKey("beginning").dataType(jDouble.getClass).cardinality(Cardinality.SINGLE).make()
//building the index
mgmt.buildIndex("journeyBeginning",classOf[Vertex]).addKey(beginning).buildCompositeIndex()
//commiting the janusGraphManagement Object 
mgmt.commit()
//reopening the janusGraphManagement Object 
mgmt: JanusGraphManagement = graph.openManagement()
//rolling back any potential transaction 
graph.tx().rollback()
//reindexing the index
mgmt.updateIndex(mgmt.getGraphIndex("journeyBeginning"),SchemaAction.REINDEX).get()
//commiting the  janusGraphManagement Object again
mgmt.commit()

I know that my index is working because when I enter this query in the gremlin console: 
:> g.V().has("beginning", 1.522135450632E12)
I get instant response

but when I do (the query which should return me the vertices I want): 
:> g.V().has("beginning",inside(1522060884498,1522363884498))
it goes over the timeout (which I set to about 10 minutes)

I think I'm doing something wrong, but I have no idea what, or where. This query is crucial to the rest of my code. 

PS: 
:> g.V().has("beginning",inside(1522060884498,1522363884498)).next(1)          OR
:> g.V().has("beginning",gte(1522060884498)).next(1) 
work fine. 
:> g.V().has("beginning",inside(1522060884498,1522363884498)).next(100) 
:> g.V().has("beginning",gte(1522060884498)).next(100) 
work as well but takes longer (and I can't limit the number of vertices I get anyway).

**Comment by twilmes** [comment]
@johanPCQLMB this looks like a bug, but can you try something for me quick and replace `inside` with `between`? I know they are slightly different so you'll need to adjust the beginning and end of your range to produce the same sized result set. Please let me know if that query with `between` runs quickly.

**Comment by johanPCQLMB** [comment]
_gremlin> :> g.V().has("beginning",between(1522060884498,1522363884498)).next(1)
==>v[9647534104]_ **(pretty much instant)**

_gremlin> :> g.V().has("beginning",between(1522060884498,1522363884498))
Script evaluation exceeded the configured 'scriptEvaluationTimeout' threshold of 500000 ms or evaluation was otherwise cancelled directly for request [g.V().has("beginning",between(1522060884498,1522363884498))]
Type ':help' or ':h' for help.
Display stack trace? [yN]_

I don't really understand the difference between between and inside, but as this is a test query I don't think changing the parameters of the query will matter a lot.

**Comment by twilmes** [comment]
@johanPCQLMB I missed this on the first read through but you're creating a composite index on `beginning`. Composite indexes do not use Elasticsearch and only support equality lookups. You'll need to create a mixed index to make use of Elasticsearch: http://docs.janusgraph.org/latest/indexes.html#index-mixed.

**Comment by johanPCQLMB** [comment]
@twilmes Thank you for your answer!

Unfortunately, my problem persists. 
I have just recreated my DB with mixed indexes : here is the code I use to create them (I have 8 indexes like this) : 
_var mgmt: JanusGraphManagement = graph.openManagement()
val beginning = mgmt.makePropertyKey("beginning").dataType(jDouble.getClass).cardinality(Cardinality.SINGLE).make()
mgmt.buildIndex("journeyBeginning",classOf[Vertex]).addKey(beginning).buildMixedIndex("search")
mgmt .commit()
mgmt: JanusGraphManagement = graph.openManagement()
 mgmt.updateIndex(mgmt.getGraphIndex("journeyBeginning"),SchemaAction.REINDEX).get()
mgmt .commit()_

I then start integrating nodes and links. 
I believe I'm not doing the index creation right, do you see anything wrong? 
Do I need to reindex after the node and edge creation? as there are millions of elements it would take a long time to do this...

**Comment by johanPCQLMB** [comment]
Hello, 

After looking into it, I have found the solution to my problem. I didn't wait for the indexes to become available before reindexing them. I did this because the method awaitGraphIndexStatus does not seem to be available in SCALA. To overcome this, I started inserting data for some time before I started reindexing. (takes a few seconds to return several thousand vertices now). 

@twilmes Thank you for your help, I think this issue can be closed.
